src/System.Text.Json/src/System/Text/Json/JsonDocument.CustomDb.cs

```diff
@@ -0,0 +1,238 @@
+        //
+        // Value types (String, Number, True, False, Null, PropertyName)
+        // * First int
+        //   * Top bit is unassigned / always clear
+        //   * 31 bits for token offset
+        // * Second int
+        //   * Top bit is unassigned / always clear
+        //   * 31 bits for the token length
+        // * Third int
+        //   * 4 bits JsonTokenType
+        //   * 28 bits unassigned / always clear
+        //
+        // EndObject / EndArray
+        // * First int
+        //   * Top bit is unassigned / always clear
+        //   * 31 bits for token offset
+        // * Second int
+        //   * Top bit is unassigned / always clear
+        //   * 31 bits for the token length (always 1, effectively unassigned)
+        // * Third int
+        //   * 4 bits JsonTokenType
+        //   * 28 bits for the number of rows until the previous value (never 0)
+        //
+        // StartObject
+        // * First int
+        //   * Top bit is unassigned / always clear
+        //   * 31 bits for token offset
+        // * Second int
+        //   * Top bit is unassigned / always clear
+        //   * 31 bits for the token length (always 1, effectively unassigned)
+        // * Third int
+        //   * 4 bits JsonTokenType
+        //   * 28 bits for the number of rows until the next value (never 0)
+        //
+        // StartArray
+        // * First int
+        //   * Top bit is unassigned / always clear
+        //   * 31 bits for token offset
+        // * Second int
+        //   * Top bit is set if the array contains other arrays or objects ("complex" types)
+        //   * 31 bits for the number of elements in this array
+        // * Third int
+        //   * 4 bits JsonTokenType
+        //   * 28 bits for the number of rows until the next value (never 0)
+        private struct CustomDb : IDisposable
+        {
+            private const int SizeOrLengthOffset = 4;
+            private const int NumberOfRowsOffset = 8;
+
+            internal int Length;
+            private byte[] _rentedBuffer;
+
+            internal CustomDb(int payloadLength)
+            {
+                // Assume that a token happens approximately every 12 bytes.
+                // int estimatedTokens = payloadLength / 12
+                // now acknowledge that the number of bytes we need per token is 12.
+                // So that's just the payload length.
+                //
+                // Add one token's worth of data just because.
+                int initialSize = DbRow.Size + payloadLength;
+
+                // Stick with ArrayPool's rent/return range if it looks feasible.
+                // If it's wrong, we'll just grow and copy as we would if the tokens
+                // were more frequent anyways.
+                const int OneMegabyte = 1024 * 1024;
+                
+                if (initialSize > OneMegabyte && initialSize <= 4 * OneMegabyte)
```

Comment:
Why is the range bounded to **4 times** 1 MB? Is this based on some relationship between the maximum initial size required and for any token size (like the assumption: `Assume that a token happens approximately every 12 bytes`) above)?

Reply:
If the heuristic says more than 4MB then 1MB is probably going to be a resize, so don't bother trying to compress it.  (It's possible that the heuristic is off by more than 4x, but that seems pretty degenerate).  This is the "if it looks feasible" part of the comment.